### ISSUE TITLE
`aws.lambda` `check-permissions` filter fails if a function's role has been deleted

### ISSUE THREAD

**Issue by ajkerrigan** [cloud-custodian/cloud-custodian#6963]
### Describe the bug

When running a policy against the `aws.lambda` resource using the `check-permissions` filter, the run errors out if a function's execution role has been deleted.

### What did you expect to happen?

I expect that the `check-permissions` filter will skip (and probably warn) if the role cannot be found, but that the policy will be able to complete successfully.

### Cloud Provider

Amazon Web Services (AWS)

### Cloud Custodian version and dependency information

```shell
Please copy/paste the following info along with any bug reports:

Custodian:   0.9.14
Python:      3.9.7 (default, Sep 30 2021, 13:53:37) 
             [GCC 10.3.0]
Platform:    posix.uname_result(sysname='Linux', nodename='mosa', release='5.14.10-051410-generic', version='#202110071109-Ubuntu SMP Thu Oct 7 14:18:05 UTC 2021', machine='x86_64')
Using venv:  True
Docker: False
Installed: 

PyJWT==1.7.1
PyYAML==5.4.1
adal==1.2.7
appdirs==1.4.4
applicationinsights==0.11.10
apscheduler==3.7.0
argcomplete==1.12.3
attrs==21.2.0
azure-cli-core==2.19.1
azure-cli-telemetry==1.0.6
azure-common==1.1.27
azure-core==1.18.0
azure-cosmos==3.2.0
azure-cosmosdb-nspkg==2.0.2
azure-cosmosdb-table==1.0.6
azure-functions==1.7.2
azure-graphrbac==0.61.1
azure-keyvault==4.1.0
azure-keyvault-certificates==4.3.0
azure-keyvault-keys==4.4.0
azure-keyvault-secrets==4.3.0
azure-mgmt-apimanagement==1.0.0
azure-mgmt-applicationinsights==1.0.0
azure-mgmt-authorization==1.0.0
azure-mgmt-batch==15.0.0
azure-mgmt-cdn==10.0.0
azure-mgmt-cognitiveservices==11.0.0
azure-mgmt-compute==19.0.0
azure-mgmt-containerinstance==7.0.0
azure-mgmt-containerregistry==8.0.0b1
azure-mgmt-containerservice==15.1.0
azure-mgmt-core==1.3.0
azure-mgmt-cosmosdb==6.4.0
azure-mgmt-costmanagement==1.0.0
azure-mgmt-databricks==1.0.0b1
azure-mgmt-datafactory==1.1.0
azure-mgmt-datalake-nspkg==3.0.1
azure-mgmt-datalake-store==1.0.0
azure-mgmt-dns==8.0.0b1
azure-mgmt-eventgrid==8.0.0
azure-mgmt-eventhub==8.0.0
azure-mgmt-hdinsight==7.0.0
azure-mgmt-iothub==1.0.0
azure-mgmt-keyvault==8.0.0
azure-mgmt-logic==9.0.0
azure-mgmt-managementgroups==1.0.0b1
azure-mgmt-monitor==2.0.0
azure-mgmt-network==17.1.0
azure-mgmt-nspkg==3.0.2
azure-mgmt-policyinsights==1.0.0
azure-mgmt-rdbms==8.1.0
azure-mgmt-redis==12.0.0
azure-mgmt-resource==16.1.0
azure-mgmt-resourcegraph==7.0.0
azure-mgmt-search==8.0.0
azure-mgmt-sql==1.0.0
azure-mgmt-storage==17.1.0
azure-mgmt-subscription==1.0.0
azure-mgmt-web==2.0.0
azure-nspkg==3.0.2
azure-storage-blob==12.9.0
azure-storage-common==2.1.0
azure-storage-file==2.1.0
azure-storage-queue==12.1.6
bcrypt==3.2.0
boto3==1.18.46
botocore==1.21.46
cachetools==4.2.2
certifi==2021.5.30
cffi==1.14.6
chardet==4.0.0
charset-normalizer==2.0.6
click==8.0.1
colorama==0.4.4
cryptography==3.4.8
decorator==5.1.0
distlib==0.3.3
dogpile.cache==1.1.4
google-api-core==2.0.1
google-api-python-client==2.22.0
google-auth==2.1.0
google-auth-httplib2==0.1.0
google-cloud-appengine-logging==0.1.4
google-cloud-audit-log==0.1.1
google-cloud-core==2.0.0
google-cloud-logging==2.6.0
google-cloud-monitoring==2.5.0
google-cloud-storage==1.42.2
google-crc32c==1.2.0
google-resumable-media==2.0.3
googleapis-common-protos==1.53.0
httplib2==0.19.1
humanfriendly==9.1
idna==3.2
importlib-metadata==4.8.1
iso8601==0.1.16
isodate==0.6.0
jmespath==0.10.0
jsonpatch==1.32
jsonpickle==1.3
jsonpointer==2.1
jsonschema==3.2.0
keystoneauth1==4.4.0
knack==0.8.0rc2
kubernetes==10.0.1
mock==4.0.3
msal==1.14.0
msal-extensions==0.3.0
msrest==0.6.21
msrestazure==0.6.4
munch==2.5.0
netaddr==0.7.20
netifaces==0.11.0
oauthlib==3.1.1
openstacksdk==0.52.0
os-service-types==1.7.0
packaging==21.0
paramiko==2.7.2
pbr==5.6.0
pkginfo==1.7.1
portalocker==1.7.1
proto-plus==1.19.0
protobuf==3.18.0
psutil==5.8.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.20
pygments==2.10.0
pynacl==1.4.0
pyopenssl==20.0.1
pyparsing==2.4.7
pyrsistent==0.18.0
python-dateutil==2.8.2
pytz==2021.1
pyyaml==5.4.1
ratelimiter==1.2.0.post0
requests==2.26.0
requests-oauthlib==1.3.0
requestsexceptions==1.4.0
retrying==1.3.3
rsa==4.7.2
s3transfer==0.5.0
setuptools==57.4.0
six==1.16.0
stevedore==3.4.0
tabulate==0.8.9
typing-extensions==3.10.0.2
tzlocal==2.1
uritemplate==3.0.1
urllib3==1.26.7
websocket-client==1.2.1
zipp==3.5.0
```

### Policy

```shell
policies:
  - name: lambda-check-permissions
    resource: aws.lambda
    filters:
      - type: check-permissions
        match: allowed
        actions:
          - "*:*"
```

### Relevant log/traceback output

```shell
2021-10-26 22:44:32,146: custodian.commands:DEBUG Loaded file /home/aj/code/sandbox-policies/lambda-xaccount.yml. Contains 1 policies
2021-10-26 22:44:32,186: custodian.aws:DEBUG using default region:us-east-2 from boto
2021-10-26 22:44:32,576: custodian.output:DEBUG Storing output with <LogFile file:///home/aj/c7n-out/lambda-check-permissions/custodian-run.log>
2021-10-26 22:44:32,620: custodian.policy:DEBUG Running policy:lambda-check-permissions resource:lambda region:us-east-2 c7n:0.9.14
2021-10-26 22:44:35,102: custodian.output:DEBUG metric:PolicyException Count:1 policy:lambda-check-permissions restype:lambda
2021-10-26 22:44:35,103: custodian.output:DEBUG metric:ApiCalls Count:8 policy:lambda-check-permissions restype:lambda
2021-10-26 22:44:35,103: custodian.output:ERROR Error while executing policy
Traceback (most recent call last):
  File "/home/aj/code/cloud-custodian/c7n/policy.py", line 290, in run
    resources = self.policy.resource_manager.resources()
  File "/home/aj/code/cloud-custodian/c7n/query.py", line 531, in resources
    resources = self.filter_resources(resources)
  File "/home/aj/code/cloud-custodian/c7n/manager.py", line 111, in filter_resources
    resources = f.process(resources, event)
  File "/home/aj/code/cloud-custodian/c7n/resources/iam.py", line 715, in process
    evaluations = self.get_evaluations(client, arn, r, actions)
  File "/home/aj/code/cloud-custodian/c7n/resources/iam.py", line 795, in get_evaluations
    evaluations = self.manager.retry(
  File "/home/aj/code/cloud-custodian/c7n/utils.py", line 446, in _retry
    return func(*args, **kw)
  File "/home/aj/.pyenv/versions/c7n-dev/lib/python3.9/site-packages/botocore/client.py", line 386, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/aj/.pyenv/versions/c7n-dev/lib/python3.9/site-packages/botocore/client.py", line 705, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.NoSuchEntityException: An error occurred (NoSuchEntity) when calling the SimulatePrincipalPolicy operation: Unable to retrieve specified IAM entity.
2021-10-26 22:44:35,110: custodian.commands:ERROR Error while executing policy lambda-check-permissions, continuing
Traceback (most recent call last):
  File "/home/aj/code/cloud-custodian/c7n/commands.py", line 301, in run
    policy()
  File "/home/aj/code/cloud-custodian/c7n/policy.py", line 1204, in __call__
    resources = PullMode(self).run()
  File "/home/aj/code/cloud-custodian/c7n/policy.py", line 290, in run
    resources = self.policy.resource_manager.resources()
  File "/home/aj/code/cloud-custodian/c7n/query.py", line 531, in resources
    resources = self.filter_resources(resources)
  File "/home/aj/code/cloud-custodian/c7n/manager.py", line 111, in filter_resources
    resources = f.process(resources, event)
  File "/home/aj/code/cloud-custodian/c7n/resources/iam.py", line 715, in process
    evaluations = self.get_evaluations(client, arn, r, actions)
  File "/home/aj/code/cloud-custodian/c7n/resources/iam.py", line 795, in get_evaluations
    evaluations = self.manager.retry(
  File "/home/aj/code/cloud-custodian/c7n/utils.py", line 446, in _retry
    return func(*args, **kw)
  File "/home/aj/.pyenv/versions/c7n-dev/lib/python3.9/site-packages/botocore/client.py", line 386, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/aj/.pyenv/versions/c7n-dev/lib/python3.9/site-packages/botocore/client.py", line 705, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.NoSuchEntityException: An error occurred (NoSuchEntity) when calling the SimulatePrincipalPolicy operation: Unable to retrieve specified IAM entity.
2021-10-26 22:44:35,114: custodian.commands:ERROR The following policies had errors while executing
 - lambda-check-permissions
```

### Extra information or context

The source of this error is [here](https://github.com/cloud-custodian/cloud-custodian/blob/0cd32854a508808333b75954c5f2fe0a45654a8e/c7n/resources/iam.py#L782-L785), in this call:

```
evaluations = self.manager.retry(
    client.simulate_custom_policy,
    PolicyInputList=[json.dumps(policy['Document'])],
    ActionNames=actions).get('EvaluationResults', ())
```

We may be able to catch `NoSuchEntityException` here and log a warning.

**Comment by kapilt** [comment]
this implies both handling for this specific case and perhaps an invalid filter for lambda if its weakly referencing role assignment.

**Comment by ajkerrigan** [comment]
The core issue here is resolved by #6986. Splitting the `invalid` enhancement request into its own issue (#6990).
